### PR TITLE
rewrite the launchable app pulldown and run button code

### PR DESCRIPTION
**Pull request by devoncarew**
@danrubel, re-write our app pulldown and launch selection code; this PRs slightly large and hits many files -

- show all existing launch configurations for the workspace
- also show any launches specific to the current file
- the currently selected launch is sticky - we don't change it as the active editor changes
- simply launch logic to only detect based on the contents of the current file (`main(`), and not use the analysis server to find all launchable resources for the project
- show the devices pulldown if the currently selected launch type supports it
- increase the width of the pulldown; the launch names now include the project name
- make the run button and cmd-R both behave the same - they both will launch what's indicated in the app pulldown

In a follow-on PR, I'll update to temporarily change the pulldown selection if the current file is runnable. Else it'll default to to the last user selection, or the last launched file if there was no selection.

### CODE REVIEW

File: lib/flutter/flutter_launch.dart [hunk 1]
```diff
@@ -30,7 +30,9 @@ class FlutterLaunchType extends LaunchType {
 
   String get flutterRunCommand => 'run';
 
-  bool canLaunch(String path) {
+  bool canLaunch(String path, { LaunchData data }) {
+    if (data == null) return false;
```
Comment (danrubel) [inline lib/flutter/flutter_launch.dart]: If data is required to launch, then why make the argument optional?

Comment (danrubel) [inline lib/flutter/flutter_launch.dart]: Ah! I found it. Other implementations don't need it. All good.

File: lib/flutter/flutter_launch.dart [hunk 2]
```diff
@@ -38,18 +40,7 @@ class FlutterLaunchType extends LaunchType {
     if (!analysisServer.isExecutable(path)) return false;
```
Comment (danrubel) [inline lib/flutter/flutter_launch.dart]: Should this line be removed?

File: lib/launch/launch_configs.dart [hunk 3]
```diff
@@ -38,6 +42,12 @@ class LaunchConfigurationManager implements Disposable, StateStorable {
     return _getCreateProjectConfig(path).getConfigs();
   }
 
+  List<LaunchConfiguration> getAllConfigs() {
```
Comment (danrubel) [inline lib/launch/launch_configs.dart]: What's the naming convention for things like getters?
Should we be using this pattern or `List<LaunchConfigurations> get allConfigs {` ?

**Comment by danrubel** [comment]
LGTM

File: lib/flutter/flutter_launch.dart (hunk 1 above)
Comment (devoncarew) [inline lib/flutter/flutter_launch.dart]: This was mostly an artifact of the refactoring to get here; not all clients use it but I can make it a normal param in any case.

File: lib/flutter/flutter_launch.dart (hunk 2 above)
Comment (devoncarew) [inline lib/flutter/flutter_launch.dart]: yup! will remove

File: lib/launch/launch_configs.dart (hunk 3 above)
Comment (devoncarew) [inline lib/launch/launch_configs.dart]: I think `get` for things that are inexpensive to call, like O(1) amortized. This could be a `get`ter as well -

**Pull request merged**